### ISSUE TITLE
How to set the in channels and out channels

### ISSUE THREAD

**Issue by dajunC** [wolny/pytorch-3dunet#16]
Hi，
Thank you for your help in this issues. Your work has helped me a lot in my field of study.

At present, I encounter such a problem.My data is a medical CT data set.In this data set, there are 50 patients. Each patient has 200 images,  and the size is 256*256. Each of the image is a single-channel grayscale image. So I set the set the raw image matrix in h5 to [50, 200, 256, 256], and label matrix is [50, 200, 256, 256]. But how can I set the 'in_channels' and 'out_channels' in the train _config.yaml?

Thank you for your answer.

**Comment by wolny** [comment]
Hi @dajunC, 

happy to hear you found this repo useful for your work.
Regarding your dataset: since you have 50 patients, those should be put in separate h5 files, or in separate datasets inside a single h5 file. Each file/internal dataset would contain [200, 256, 256] raw image. Creating a single h5 dataset of [50, 200, 256, 256] is not what you want to make it work, since that shape would imply that you have 50 input channels to the network which is incorrect in your case.
Since you've mentioned that slices are 'single-channel grayscale' then the `in_channels` should be set to `1` (see: https://github.com/wolny/pytorch-3dunet/blob/master/resources/train_config.yaml#L6). The number of output channels depends on the number of different labels you have in the ground truth.

Good luck.

**Comment by dajunC** [comment]
Thank you for your help.

That means I should put one patients in the h5, and the size is 200*256*256. But, how can I train all the patients?

**Comment by wolny** [comment]
Just state which files you want to use for training in the config file (https://github.com/wolny/pytorch-3dunet/blob/master/resources/train_config.yaml#L69). E.g.
```
train_path:
    - 'patient_1.h5'
    - 'patient_2.h5'
    ...
    - 'patient_n.h5'
```

**Comment by dajunC** [comment]
I use your method and get a error.

`2019-04-17 20:52:01,383 [MainThread] INFO UNet3DTrainer - Number of learnable params 4082082
2019-04-17 20:52:02,717 [MainThread] INFO UNet3DTrainer - Sending the model to 'cuda:0'
2019-04-17 20:52:02,737 [MainThread] INFO UNet3DTrainer - Training iteration 1. Batch 0. Epoch [0/49]
2019-04-17 20:52:03,874 [MainThread] INFO UNet3DTrainer - Training iteration 2. Batch 1. Epoch [0/49]
/opt/conda/conda-bld/pytorch_1533739672741/work/aten/src/THCUNN/SpatialClassNLLCriterion.cu:99: void cunn_SpatialClassNLLCriterion_updateOutput_kernel(T *, T *, T *, long *, T *, int, int, int, int, int, long) [with T = float, AccumT = float]: block: [0,0,0], thread: [319,0,0] Assertion `t >= 0 && t < n_classes` failed.`

**Comment by wolny** [comment]
This error indicates that the number of output channels from the network `out_channels` does not correspond to the number of labels you have in the ground truth (see e.g.g https://discuss.pytorch.org/t/is-there-anybody-happen-this-error/17416). Make sure they match and you're good to go.

**Comment by dajunC** [comment]
Thank you every much. I have work it. Thank you for your help
